clients/src/test/java/org/apache/kafka/clients/producer/internals/ProducerBatchTest.java

```diff
@@ -70,4 +73,14 @@ public void testLargeFullOldNowExpire() {
         // Set `now` to 2ms before the create time.
         assertFalse(batch.maybeExpire(10240, 10240L, now - 2L, 10240L, true));
     }
+
+    @Test
+    public void testShouldNotAppendOnceRecordBatchGetsFull() {
+        ProducerBatch batch = new ProducerBatch(new TopicPartition("topic", 1), memoryRecordsBuilder, now);
+        FutureRecordMetadata result0 = batch.tryAppend(now, null, new byte[10], Record.EMPTY_HEADERS, null, now);
+        assertNotNull(result0);
+        memoryRecordsBuilder.closeForRecordAppends();
+        assertFalse(memoryRecordsBuilder.hasRoomFor(now, null, new byte[10]));
```

Comment:
I'd add this same test (but `assertTrue`) just before `closeForRecordAppends` to verify that the `closeForRecordApends` is what causes the change.